src/engine/GroupBy.cpp

```diff
@@ -801,76 +794,71 @@ bool GroupBy::computeGroupByForSortedSubtree(IdTable* result,
-  auto numBlocks = (size_t)ceil(((double)subresult.size()) / blockSize);
 
-  size_t rowIndex = 0;
-
-  for (size_t i = 0; i < numBlocks; ++i) {
-    evaluationContext._beginIndex = i * blockSize;
-    evaluationContext._endIndex =
-        std::min((i + 1) * blockSize, subresult.size());
+  for (size_t i = 0; i < subresult.size(); i += blockSize) {
+    evaluationContext._beginIndex = i;
+    evaluationContext._endIndex = std::min(i + blockSize, subresult.size());
 
     auto currentBlockSize =
         evaluationContext._endIndex - evaluationContext._beginIndex;
 
-    size_t aggregateIndex = 0;
-    auto currentRowIndex = rowIndex;
-
     // Perform HashMap lookup for all groups in current block
-    std::vector<std::array<AverageAggregationData, numAggregates>*> hashEntrys;
+    std::vector<std::array<AverageAggregationData, numAggregates>*> hashEntries;
+
+    auto getId = [&subresult, &evaluationContext, &columnIndex](size_t j) {
+      return subresult(evaluationContext._beginIndex + j, columnIndex);
+    };
 
     // Create elements in map
     for (size_t j = 0; j < currentBlockSize; ++j) {
-      auto id = subresult(evaluationContext._beginIndex + j, columnIndex);
+      auto id = getId(j);
       (void)map[id];
     }
 
     // Get pointers to values
     // (valid for this iteration as no new elements added)
     for (size_t j = 0; j < currentBlockSize; ++j) {
-      auto id = subresult(evaluationContext._beginIndex + j, columnIndex);
-      hashEntrys.emplace_back(&(map.at(id)));
+      auto id = getId(j);
+      hashEntries.emplace_back(&(map.at(id)));
     }
 
+    // TODO<C++23> use views::enumerate
+    size_t aggregateIndex = 0;
     for (auto& aggregate : aggregates) {
       // Evaluate child expression on block
       auto expr = aggregate._expression.getPimpl();
       auto exprChildren = expr->children();
       sparqlExpression::ExpressionResult expressionResult =
           exprChildren[0]->evaluate(&evaluationContext);
 
-      auto visitor = [&]<sparqlExpression::SingleExpressionResult T>(
-                         T&& singleResult) mutable {
-        auto generator = sparqlExpression::detail::makeGenerator(
-            std::forward<T>(singleResult), currentBlockSize,
-            &evaluationContext);
+      auto visitor =
+          [&currentBlockSize, &evaluationContext, &i, &hashEntries,
+           &aggregateIndex]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) mutable {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), currentBlockSize,
+                &evaluationContext);
 
-        using NVG = sparqlExpression::detail::NumericValueGetter;
+            using NVG = sparqlExpression::detail::NumericValueGetter;
 
-        for (const auto& val : generator) {
-          sparqlExpression::detail::NumericValue numVal =
-              NVG()(val, &evaluationContext);
+            auto rowIndex = i;
```

Comment:
This line looks like `i` should be called `startOfBlockIdx` or something similar.
Or... At a closer look
simply write `rowIndex = 0` then you don't need to subtract the exact same value (but from a different value) again three lines down.